src/jit.c

```diff
@@ -0,0 +1,1557 @@
+                                   int rn,
+                                   int32_t imm7)
+{
+    int32_t imm_div = ((op == LSP_STPX) || (op == LSP_LDPX)) ? 8 : 4;
+    assert(imm7 % imm_div == 0);
+    imm7 /= imm_div;
+    emit_a64(state, op | (imm7 << 15) | (rt2 << 10) | (rn << 5) | rt);
+}
+
+/* [ARM-A]: C4.1.65: Unconditional branch (register).  */
+static void emit_uncond_branch_reg(struct jit_state *state,
+                                   a64opcode_t op,
+                                   int rn)
+{
+    emit_a64(state, op | (rn << 5));
+}
+
+/* [ARM-A]: C4.1.67: Data-processing (2 source).  */
+static void emit_dataproc_2source(struct jit_state *state,
+                                  bool is64,
+                                  a64opcode_t op,
+                                  int rd,
+                                  int rn,
+                                  int rm)
+{
+    emit_a64(state, sz(is64) | op | (rm << 16) | (rn << 5) | rd);
+}
+
+
+/* [ARM-A]: C4.1.67: Data-processing (3 source).  */
+static void emit_dataproc_3source(struct jit_state *state,
+                                  bool is64,
+                                  a64opcode_t op,
+                                  int rd,
+                                  int rn,
+                                  int rm,
+                                  int ra)
+{
+    emit_a64(state, sz(is64) | op | (rm << 16) | (ra << 10) | (rn << 5) | rd);
+}
+
+static void update_branch_imm(struct jit_state *state,
+                              uint32_t offset,
+                              int32_t imm)
+{
+    assert((imm & 3) == 0);
+    uint32_t insn;
+    imm >>= 2;
+    memcpy(&insn, state->buf + offset, sizeof(uint32_t));
+    if ((insn & 0xfe000000U) == 0x54000000U /* Conditional branch immediate. */
+        || (insn & 0x7e000000U) ==
+               0x34000000U) { /* Compare and branch immediate.  */
+        assert((imm >> 19) == INT64_C(-1) || (imm >> 19) == 0);
+        insn |= (imm & 0x7ffff) << 5;
+    } else if ((insn & 0x7c000000U) == 0x14000000U) {
+        /* Unconditional branch immediate.  */
+        assert((imm >> 26) == INT64_C(-1) || (imm >> 26) == 0);
+        insn |= (imm & 0x03ffffffU) << 0;
+    } else {
+        assert(false);
+        insn = BAD_OPCODE;
+    }
+#if defined(__APPLE__) && defined(__aarch64__)
+    pthread_jit_write_protect_np(false);
+#endif
+    memcpy(state->buf + offset, &insn, sizeof(uint32_t));
+#if defined(__APPLE__) && defined(__aarch64__)
+    pthread_jit_write_protect_np(true);
+#endif
+}
+#endif
+
+static inline void emit_jump_target_offset(struct jit_state *state,
+                                           uint32_t jump_loc,
+                                           uint32_t jump_state_offset)
+{
+    struct jump *jump = &state->jumps[state->n_jumps++];
+    jump->offset_loc = jump_loc;
+    jump->target_offset = jump_state_offset;
+}
+
+/* The REX prefix and ModRM byte are emitted.
```

Comment:
This comment should only make sense to x64 backend. Move into the function body.